### DISCUSSION TITLE
Should the be filed as a MPS bug, a bug on torchmetrics, or something else? Negative average precision

### DISCUSSION THREAD

**Discussion by el-hult** [Lightning-AI/torchmetrics#3276] [Q&A]
Hello. Consider the code below, running on a MacBrook Pro with M1 mps accelleration.

```
import pathlib
import torch
from torchmetrics.classification import MultilabelAveragePrecision

# load a tensor of strange logits and targets
with open(
  pathlib.Path(__file__).parent / "strange.pt",
  "rb",
) as f:
  foo = torch.load(f)

logits = foo["logits"]
target = foo["target"]

for device, determinism in [
  ("mps", False),
  ("mps", True),
  ("cpu", True),
]:
  torch.use_deterministic_algorithms(determinism)
  metrics = MultilabelAveragePrecision(num_labels=10, average="micro").to(device)
  ap = metrics(preds=logits.to(device), target=target.to(device))
  print(device, determinism, ap)

```

with the supplied data file. The output on my machine is 

```
mps False tensor(-334908.7500, device='mps:0')
mps True tensor(nan, device='mps:0')
cpu True tensor(0.6495)
```

The data is logits and targets from one hot encoded MNIST and a poorly fitted logistic regression.
If you subset the data and take a few thousand subsamples, it works well. but for some reasons, the results are crazy on the MPS if nondeterminism is allowed. And without it, is is still super strange to get a 'nan'. This should really have a finite answer all the time. Which the cpu computation gives. 

I'd be happy to supply the pt file, but im not allowed to attach it.
